lib/tdf3/src/tdf.ts

```diff
@@ -125,6 +128,97 @@ type TDFConfiguration = {
   cryptoService: CryptoService;
 };
 
+export type RewrapRequest = {
+  signedRequestToken: string;
+};
+
+export type KasPublicKeyInfo = {
+  url: string;
+  algorithm: KasPublicKeyAlgorithm;
+  kid?: string;
+  pem: string;
+};
+
+export type KasPublicKeyAlgorithm = 'ec:secp256r1' | 'rsa:2048';
+
+export type KasPublicKeyFormat = 'pkcs8' | 'jwks';
+
+type KasPublicKeyParams = {
+  algorithm?: KasPublicKeyAlgorithm;
+  fmt?: KasPublicKeyFormat;
+  v?: '1' | '2';
+};
+
+export type RewrapResponse = {
+  entityWrappedKey: string;
+  sessionPublicKey: string;
+};
+
+/**
+ * If we have KAS url but not public key we can fetch it from KAS, fetching
+ * the value from `${kas}/kas_public_key`.
+ */
+export async function fetchKasPublicKey(
+  kas: string,
+  strict = true,
+  algorithm?: KasPublicKeyAlgorithm
+): Promise<KasPublicKeyInfo> {
+  if (!kas) {
+    throw new TdfError('KAS definition not found');
+  }
+  validateSecureUrl(kas, strict);
+  const infoStatic = { url: kas, algorithm: algorithm || 'rsa:2048' };
+  const params: KasPublicKeyParams = {};
+  if (algorithm) {
+    params.algorithm = algorithm;
+  }
+  try {
+    const response: { data: string | KasPublicKeyInfo } = await axios.get(`${kas}/kas_public_key`, {
+      params: {
+        ...params,
+        v: '2',
+      },
+    });
+    const pem =
+      typeof response.data === 'string'
+        ? await TDF.extractPemFromKeyString(response.data)
+        : response.data.pem;
+    return {
+      pem,
+      ...infoStatic,
+      ...(typeof response.data !== 'string' && response.data.kid && { kid: response.data.kid }),
+    };
+  } catch (cause) {
+    if (cause?.response?.status != 400) {
+      throw new TdfError(
+        `Retrieving KAS public key [${kas}] failed [${cause.name}] [${cause.message}]`,
+        cause
+      );
+    }
+  }
+  // Retry with v1 params
```

Comment:
it will get here if any exception is thrown unless the exception has a response code of 400?